tools/apps/src/lib.rs

```diff
@@ -0,0 +1,917 @@
+    ) {
+        let mut reqs_done = 0;
+
+        for req in self.reqs.iter_mut().skip(self.reqs_sent) {
+            info!("sending HTTP request {:?}", req.hdrs);
+
+            let s = match self.h3_conn.send_request(
+                conn,
+                &req.hdrs,
+                self.body.is_none(),
+            ) {
+                Ok(v) => v,
+
+                Err(quiche::h3::Error::TransportError(
+                    quiche::Error::StreamLimit,
+                )) => {
+                    debug!("not enough stream credits, retry later...");
+                    break;
+                },
+
+                Err(e) => {
+                    error!("failed to send request {:?}", e);
+                    break;
+                },
+            };
+
+            req.stream_id = Some(s);
+            req.response_writer =
+                make_writer(&req.url, target_path, req.cardinal);
+
+            if let Some(body) = &self.body {
+                if let Err(e) = self.h3_conn.send_body(conn, s, body, true) {
+                    error!("failed to send request body {:?}", e);
+                    break;
+                }
+            }
+
+            reqs_done += 1;
+        }
+
+        self.reqs_sent += reqs_done;
+    }
+
+    fn handle_responses(
+        &mut self, conn: &mut quiche::Connection, buf: &mut [u8],
+        req_start: &std::time::Instant,
+    ) {
+        loop {
+            match self.h3_conn.poll(conn) {
+                Ok((stream_id, quiche::h3::Event::Headers { list, .. })) => {
+                    info!(
+                        "got response headers {:?} on stream id {}",
+                        list, stream_id
+                    );
+                },
+
+                Ok((stream_id, quiche::h3::Event::Data)) => {
+                    if let Ok(read) = self.h3_conn.recv_body(conn, stream_id, buf)
+                    {
+                        debug!(
+                            "got {} bytes of response data on stream {}",
+                            read, stream_id
+                        );
+
+                        let req = self
+                            .reqs
+                            .iter_mut()
+                            .find(|r| r.stream_id == Some(stream_id))
+                            .unwrap();
+
+                        match &mut req.response_writer {
+                            Some(rw) => {
+                                rw.write_all(&buf[..read]).ok();
+                            },
+
+                            None => {
+                                print!("{}", unsafe {
+                                    std::str::from_utf8_unchecked(&buf[..read])
+                                });
+                            },
+                        }
+
+                        if let Some(f) = &mut req.response_writer {
+                            f.write_all(&buf[..read]).ok();
+                        }
```

Comment:
```suggestion
```

Reply:
ah yeah this is stupid, thanks for finding it